### PR TITLE
store-gateway: record request metrics for LabelValues & LabelNames calls

### PR DESCRIPTION
**Pull request by Gabrielopesantos**
#### What this PR does

Collects prometheus metrics for postings/series/chunks/etc fetching from LabelValues() and LabelNames() calls. This is already done for Series()

#### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir/issues/4595

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by dimitarvdimitrov — COMMENTED** [review]
Nice work! This looks like it's going in the right direction. The next step would be use the counters, summaries, histograms in `recordLabel(Names|Values)CallResult`.

I can help by shedding some light on what statistic is relevant for which calls (LabelValues, LabelNames, Series).

The fields that begin with `postings*` or `cachedPostings*` in `queryStats` are relevant for all 3 RPC calls, since postings are the way to find which timeseries match the request (this is sometimes not true for LabelNames without matchers)

Stats that begin with `series*` or `streamingSeries*` are relevant for LabelNames and Series only. LabelValues does not fetch series, so these values will always be 0 and observing them will throw off the summaries. `mergedSeriesCount` is only relevant for the Series RPC.

Stats that begin with `chunks*` are only relevant for the Series RPC.

I hope this didn't make things more complicated 😄

**Comment by Gabrielopesantos** [comment]
Hey,
It didn't make things more complicated at all, thank you for the help. I have committed the methods for LabelNames and LabelValues. With that I believe the next steps would be setting up some of the calculations for the metrics in both "call handlers", adding support for the new metrics that you pointed out in the issue and testing the behaviour of these additions.

File: pkg/storegateway/bucket.go [hunk 1]
```diff
@@ -993,6 +1041,12 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
+	stats.update(func(stats *queryStats) {
+		stats.blocksQueried = len(sets)
+		stats.streamingSeriesExpandPostingsDuration += time.Since(begin)
+	})
+	s.metrics.seriesBlocksQueried.Observe(float64(len(sets)))
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/bucket.go:1048]: I think this belongs better in the `record*CallResult` functions. I see it is done in the same way in the `streamingSeriesSetForBlocks` function, but I think that was a mistake. Can you move both usages to `record*CallResult`?

**Comment by dimitarvdimitrov** [comment]
> I believe the next steps should be setting up some of the calculations for the metrics in both "call handlers"

I had another look at the metrics we report, I don't think 			`streamingSeriesEncodeResponseDuration` and `streamingSeriesSendResponseDuration` make much sense for LabelValues and LabelNames. These two metrics are meant to record how much time is spent in sending the series response to the querier. Since there is no series response in the LabelValues and LabelNames RPCs, I think we shouldn't record these metrics. The Label* calls' responses are also usually much smaller, so they might throw away the averages.

I also clarified my previous answer - we don't need to report `mergedSeriesCount` in neither LabelValues nor LabelNames.

> adding support for the new metrics that you pointed out in the issue

All the metrics that are in that issue are metrics that we already report, so I don't think there' new metrics to record. What did you mean?

> testing the behaviour of these additions

Agree that testing is a good idea. Currently the metrics from Series are tested in `assertQueryStatsMetricsRecorded`, which is in turn used by `TestBucketStore_e2e`. I think adding similar assertions to `TestBucketStore_LabelNames_e2e` and `TestBucketStore_LabelValues_e2e` will be enough.

**Review by dimitarvdimitrov — COMMENTED** [review]
@Gabrielopesantos is there anything I can do to help with this PR? I was hoping to be able to use these metrics with the changes in https://github.com/grafana/mimir/issues/4593, which I'll be wrapping up this week.

What do you think about merging only the production code changes and then working on tests in a follow-up PR? The rest of the changes in this PR seem small (+ making the tests compile), so I think we can get them done soon. I'm happy to work on some of this if you don't have capacity in the coming days. Let me know!

**Comment by Gabrielopesantos** [comment]
Hey @dimitarvdimitrov 
Sorry for the delay in getting back to the PR. If there are no new metrics to be recorded (it might have been some confusion from by side), which you mention [here](https://github.com/grafana/mimir/pull/4673#issuecomment-1502030929), I believe I am done with the implementation (excluding tests).

About your suggestion, only the production code changes and then working on tests in a follow-up PR, I am ok with that as long as you are. If this PR is somehow blocking you from doing your work, feel free to take ownership of this PR and take care of what is missing or we can close it without merging.

**Comment by dimitarvdimitrov** [comment]
> About your suggestion, only the production code changes and then working on tests in a follow-up PR, I am ok with that as long as you are.

I think it's ok given that the current behaviour is already tested in `assertQueryStatsMetricsRecorded`.

> If this PR is somehow blocking you from doing your work, feel free to take ownership of this PR and take care of what is missing or we can close it without merging.

I just pushed one small commit to remove expandedPostings stage duration from LabelValues because currently it's not easy to be precisely measured in LabelNames.

File: pkg/storegateway/bucket.go [hunk 2]
```diff
@@ -980,6 +1025,11 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
+	stats.update(func(stats *queryStats) {
+		stats.blocksQueried = len(sets)
+		stats.streamingSeriesExpandPostingsDuration += time.Since(begin)
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/bucket.go:1030]: This time seems to also include the time to fetch series. I will push a commit to remove this.
```suggestion
```

File: pkg/storegateway/bucket.go [hunk 3]
```diff
@@ -915,7 +954,13 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
 		return nil, status.Error(codes.InvalidArgument, errors.Wrap(err, "translate request labels matchers").Error())
 	}
 
-	resHints := &hintspb.LabelNamesResponseHints{}
+	var (
+		stats    = newSafeQueryStats()
+		begin    = time.Now()
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/bucket.go:959]:
```suggestion
```

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM! Thank you for working on this. And sorry if I was a bit pushy, i was hoping to get this into the r234 weekly release on Monday.

Let me know if you're still interested in contributing the tests in another PR

**Comment by dimitarvdimitrov** [comment]
I also added a changelog entry and updated the metrics help texts because they mentioned Series only. 

This is now ready for a review. I would like to merge it by the end of the day. So if anyone else wants to review afterwards, please do - comments can be addressed post-merge as well.

**Comment by Gabrielopesantos** [comment]
> LGTM! Thank you for working on this. And sorry if I was a bit pushy, i was hoping to get this into the r234 weekly release on Monday.
> 
> Let me know if you're still interested in contributing the tests in another PR

Hey @dimitarvdimitrov 
No worries, thank you for the feedback and the help.

I had a look on the test, that were not included in this PR, and based on your tips above created [this](https://github.com/grafana/mimir/pull/4799). Let me know if it is aligned with what you were expecting. Thank you

**Pull request merged**